pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/ScanBasedDocIdIterator.java

```diff
@@ -34,10 +36,17 @@
  */
 public interface ScanBasedDocIdIterator extends BlockDocIdIterator {
 
+  MutableRoaringBitmap applyAnd(BatchIterator batchIterator, OptionalInt firstDoc, OptionalInt lastDoc);
```

Comment:
Trying to understand why do we need to add this API. The `BatchIterator` should always be returned from the roaring bitmap. If you need to pass `RoaringBitmap` here, you may modify the existing `applyAnd()` to take an `ImmutableBitmapDataProvider`.
In order to apply a different start/end to the bitmap, we may do an AND operation to the bitmap with the range.